pennylane/ops/op_math/composite.py

```diff
@@ -0,0 +1,226 @@
+    """
+
+    _eigs = {}  # cache eigen vectors and values like in qml.Hermitian
+    _op_symbol = None
+    _name = None
+
+    def __init__(
+        self, *operands: Operator, do_queue=True, id=None
+    ):  # pylint: disable=super-init-not-called
+        self._id = id
+        self.queue_idx = None
+
+        if self._name is None:
+            raise NotImplementedError("Child class must specify _name")
+        if self._op_symbol is None:
+            raise NotImplementedError("Child class must specify _op_symbol")
+        if len(operands) < 2:
+            raise ValueError(f"Require at least two operators to combine; got {len(operands)}")
+
+        self.operands = operands
+        self._wires = qml.wires.Wires.all_wires([op.wires for op in self.operands])
+        self._hash = None
+
+        if do_queue:
+            self.queue()
+
+    def __repr__(self):
+        return f" {self.op_symbol} ".join(
+            [f"({op})" if op.arithmetic_depth > 0 else f"{op}" for op in self.operands]
+        )
+
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_op.operands = tuple(s.__copy__() for s in self.operands)
+
+        for attr, value in vars(self).items():
+            if attr not in {"operands"}:
+                setattr(copied_op, attr, value)  # TODO: exclude data?
+
+        return copied_op
+
+    def __iter__(self):
+        """Return the iterator over the underlying operands."""
+        return iter(self.operands)
+
+    def __getitem__(self, idx):
+        """Return the operand at position ``idx`` of the composition."""
+        return self.operands[idx]
+
+    def __len__(self):
+        """Return the number of operators in this composite operator"""
+        return len(self.operands)
+
+    @property
+    def op_symbol(self) -> str:
+        """The symbol used when visualizing the composite operator"""
+        return self._op_symbol
+
+    @property
+    def data(self):
+        """Create data property"""
+        return [op.parameters for op in self.operands]
+
+    @data.setter
+    def data(self, new_data):
+        """Set the data property"""
+        for new_entry, op in zip(new_data, self.operands):
+            op.data = new_entry
+
+    @property
+    def num_wires(self):
+        return len(self.wires)
+
+    @property
+    def num_params(self):
+        return sum(op.num_params for op in self.operands)
+
+    @property
+    @abc.abstractmethod
+    def is_hermitian(self):
+        """
+        TODO: should we do `all(op.is_hermitian for op in self.operands)` as default?
+        """
+
+    @abc.abstractmethod
+    def matrix(self, wire_order=None):
+        """Representation of the operator as a matrix in the computational basis."""
+
+    @abc.abstractmethod
+    def sparse_matrix(self, wire_order=None):
+        """Compute the sparse matrix representation of the composite operator."""
```

Comment:
should we require all child classes implement this? the `SparseMatrixUndefinedError` will [come from the parent Operator class](https://github.com/PennyLaneAI/pennylane/blob/3306eddbf2822c4de958d75519d2e4a235c54dc5/pennylane/operation.py#L498-L539) anyway

Reply:
I think we can leave these representations out.  If they are not implemented, then we should just have the base `Operator` behavior.